[docs] [pipes] - Add Concept page (#17048)
dagster-io/dagster: docs/content/guides/dagster-pipes.mdx

---
title: Dagster Pipes | Dagster Docs
description: "Dagster Pipes provides a protocol between the orchestration environment (Dagster) and external execution (ex: Databricks) and a toolkit for building implementations of that protocol."
---

# Dagster Pipes

> At a high level, Dagster Pipes allows you to send Dagster context data into an external process which is then sent back to Dagster.

Reviewer: Btw I had written this is Notion as the first few paragraphs:

"Dagster Pipes is a toolkit for building integrations between Dagster and external execution environments. It standardizes the process of passing parameters, injecting context information, ingesting logs, and collecting metadata all while remaining agnostic to how those remote computations are launched in those environments. This enables the clean separation of orchestration and business logic in the Dagster ecosystem.

It also smooths the process of incorporating pre-existing code, business logic, and execution environments into Dagster. Using Dagster Pipes, you add just a few lines to you code, and your code becomes seamlessly invocable by the orchestrator, and in turn can stream logs and metadata back to Dagster so you can leverage its observability, lineage, cataloging, and debugging capabilities."

Author: thanks! I started moving this right before you added that. I updated this.

For example, let’s say you want to incorporate some business logic in a Dagster pipeline. However, it needs to deal with pre-existing data pipelines and Python environments, it’s written in a language other than Python, or you wish to complete separate the orchestration and business logic execution environment.,

Using Dagster Pipes, you can write a few lines of code in your business logic in an external environment and send metadata and event logs back to Dagster.

---

## Benefits

With Dagster Pipes, you can:

- Incorporate existing code into Dagster without huge refactors
- Onboard stakeholder teams onto Dagster incrementally

> - Run code in external environments (with minimal dependencies) and stream log and structured metadata back to Dagster

Reviewer: Run code in external environments and easily pass parameters and context to that code, and then stream unstructured logs and structured metadata back to Dagster

- Separate orchestration and business logic environments
- Use languages other than Python with Dagster

---

## Limitations

While Dagster Pipes is lightweight and flexible, there are a few limitations to be aware of:

> - **Step launchers are currently incompatible.**

Reviewer: > **Step launchers are currently incompatible.**

I don't think is an accurate framing. Pipes are a lightweight _alternative_ to step launchers.

Author: That's fair. Would you agree that it's accurate to portray this as an either-or situation - that is, you can use step launchers, and you could use pipes, but not necessarily together?

Reviewer: Definitely either-or.

- We have some external code that contains some business logic.
- Our code also calculates the number of records it processes at a given time. We want to send this metadata back to Dagster.
- We have a Dagster asset that:
  - Provides access to asset context data, such as logging, via <PyObject object="AssetExecutionContext" />

>   - Invokes the external code

Reviewer: It's important to mention parameter passing here

> - **I/O managers aren’t currently supported.** Dagster Pipes isn't built for processing data in memory, which is what I/O managers are for. For example, to process data in memory in Pyspark, you can use an I/O manager [as demonstrated in the Pyspark integration guide](/integrations/spark#running-pyspark-code-in-assets). Otherwise, in a remote case, you can use Pipes.

Reviewer: > I took another swing at this. It does sound like if you did include the core Dagster library, you could use those concepts, though - which sorta defeats the purpose of using Pipes and its lack of dependencies.

Yeah that's right. I don't think we need to even mention this possibility, even though it is theorectically possible.

Refer to the following diagram and explanation for how Dagster Pipes would accomplish this, step-by-step. **Note:** Depending on the technology you use (ex: Databricks, Docker), this implementation of the protocol may vary. Refer to the [Understanding Dagster Pipes process guide](/guides/dagster-pipes/understanding-dagster-pipes) for more info.

<Image
alt="Diagram explaining the Dagster Pipes process"

> src="/images/guides/dagster-pipes/dagster-pipes-process.png"

Reviewer 2: the resolution of the image isn't great when zoom in

width={1000}
height={393}
/>

### In the orchestration process (Dagster)

1. The Dagster asset is executed.
2. A connection to the external process, such as Databricks, is established.
3. Dagster sends context info (ex: `partition_key`, `asset_key`, etc.) and any provided parameters to the external process.

> 

Reviewer: So I think this is in a sort uncanny valley of explanation, where it is just sufficiently "in the weeds" to be confusing, because it isn't quite right.

A slightly more detailed can be more accurate:

### Orchestration process

1. The Dagster asset is executed.
2. Context information is made available to the external process through an agreed upon side channel, such as an object in s3.
3. Separate side channels for unstructured logs and structured metadata results are initialized.
4. The external process is launched from Dagster. Bootstrap parameters are passed, which will be used to load the context information from the side channel.

### In the external process
1. The process starts
2. The user calls code (`open_dagster_pipes`) which loads the context from the side channel.
3. The process proceeds, writing unstructured logs and streaming metadata back to the side channels initialized before launch.

The other option is to make it even more high level.

Reviewer 2: in the more detailed version, the term "side channel" stood out to me as yet another thing that we need to explain - mentioning it here could be overwhelming to audience like stakeholders, and therefore defeat the purpose of this concept page (e.g. pick your own journey). so im leaning towards to make it even more high level and leave the accurate explanation in the customizing protocol doc.

Reviewer 3: For 
```
1. The Dagster asset is executed.
```
The asset isn't executed to completion before step 2 starts right? that's what this line implies to me, but maybe my understanding of how pipes works is off. 

For 
```
4. The external process is launched from Dagster. Bootstrap parameters are passed, which will be used to load the context information from the side channel.
```
TIL "bootstrap" has an actual technical meaning and isn't a buzzword.   So maybe for the stakeholder audience we should use a different term?

Author: I agree with Yuhan - I think that's a bit too much for this page, which was always intended to be pretty high-level so it's consumable for most (if not all) audiences. I figure if someone wants more details, they probably want to do some customization work, and should look next at the more technical `Understanding the Dagster Pipes process` guide.

> ---

Reviewer 3: is another PR adding this to the guides landing page?

Author: Yep! Just hadn't gotten to it yet.

Reviewer 4: Will this be marked as Experimental? I don't see any mention of that on the page

Author: @PedramNavid Done!

title: Dagster Pipes | Dagster Docs
description: "Dagster Pipes provides a protocol between the orchestration environment (Dagster) and external execution (ex: Databricks) and a toolkit for building implementations of that protocol."
---

# Dagster Pipes

Dagster Pipes is a toolkit for building integrations between Dagster and external execution environments. It standardizes the process of passing parameters, injecting context information, ingesting logs, and collecting metadata all while remaining agnostic to how remote computations are launched in those environments. This enables the separation of orchestration and business logic in the Dagster ecosystem.

It also smooths the process of incorporating pre-existing code, business logic, and execution environments into Dagster. Using Dagster Pipes, you add just a few lines to your code, and it becomes seamlessly invocable by the orchestrator. In turn, you can stream logs and metadata back to Dagster so you can leverage its observability, lineage, cataloging, and debugging capabilities.

---

> ## Benefits

Reviewer 5: I think it would be helpful to clarify the answer to the question: "what is the best practice way for using Dagster?"

If the answer is more of a decision tree, we should include that decision tree

But I think it is very important that we clarify for both new dagster prospects and existing customers how pipes relates to assets w/out pipes (and maybe even just as important, our integrations w/out pipes)

In other words, a good addition would be: 
- when should I use pipes?
- when should I not?

Reviewer: I think a cohesive decision tree about overall integration usage is beyond the scope of this guide given the muddled state of our integration ecosystem. @PedramNavid and I have discussed this in the past and I think making progress on this is a good candidate for quality corner.

In the short-term, I think we should frame Pipes as a much better way of richly integrating with external execution environments that cannot or do not include the core dagster module, and where Dagster does not control the deployment of code to those external environments.

It also smooths the process of incorporating pre-existing code, business logic, and execution environments into Dagster. Using Dagster Pipes, you add just a few lines to your code and it becomes seamlessly invocable by the orchestrator. In turn, you can stream logs and metadata back to Dagster so you can leverage its observability, lineage, cataloging, and debugging capabilities.

---

## Benefits

With Dagster Pipes, you can:

- Incorporate existing code into Dagster without huge refactors
- Onboard stakeholder teams onto Dagster incrementally
- Run code in external environments, and:
  - Easily pass parameters to the code
  - Stream unstructured logs and structured metadata back to Dagster
- Separate orchestration and business logic environments
- Use languages other than Python with Dagster

---

## Limitations

While Dagster Pipes is lightweight and flexible, there are a few limitations to be aware of:

- **Step launchers are currently incompatible.**

> - **I/O managers aren’t currently supported.** If you rely heavily on [I/O managers](/concepts/io-management/io-managers) in remote execution, we recommend using step launchers.

Author: @yuhan Is this correct? It feels weird to say I/O managers aren't supported for this, so use step launchers, which also aren't supported. I feel like there's something I missed in our original meeting.

Reviewer 2: I think the io manager "incompatibility" means:

Dagster Pipes is not built for processing data in memory, which I/O managers are for. For example, if you'd like to process data in memory in Pyspark, use I/O manager as demonstrated  https://docs.dagster.io/integrations/spark#running-pyspark-code-in-assets; otherwise (in a remote case), use Pipes.

Author: 🙏

---

## How it works

Dagster Pipes provides a protocol between the orchestration environment (Dagster) and external execution (ex: Databricks) and a toolkit for building implementations of that protocol.

When Dagster Pipes is invoked, several steps will be carried out both in the **orchestration process**, or Dagster, and in the **external process**, such as Databricks.

### In the orchestration process (Dagster)

When the execution of a Dagster asset begins, Dagster makes context information (ex: `partition_key`, `asset_key`, etc.) available for the external process to consume.

> Next, Dagster launches the external process and passes in any provided parameters.

Reviewer 2: Wonder if this can be even more high level like:

"When Dagster Pipes is called in a Dagster asset, Dagster launches the external process with parameters and context information (ex: `partition_key`, `asset_key`, etc.) "

Author: That's fine by me

<Image
alt="Diagram explaining the Dagster Pipes process"
src="/images/guides/dagster-pipes/dagster-pipes-process.png"
width={1000}
height={393}
/>

### In the external process

> The process starts and loads the context info provided by Dagster. When finished, the process sends execution data, logs, and any specified metadata, back to Dagster.

Reviewer 2: > When finished, the process sends execution data, logs, and any specified metadata, back to Dagster.

This isn't 100% accurate. The information gets streamed back to Dagster through that "side channel" while the process is running, meaning the information (logs, metadata) can be sent back to Dagster even before the external process finishes.

Author: Any suggestions for how to word this? Given that we wanted to leave 'side channel' out of this page, I don't want to include it here.

Maybe?: "When the process is finished, execution data, logs, and any specified metadata will be sent back to Dagster." ?

Reviewer: It typically happens continuously while the process is executing, not just sent on completion.

Author: > The process starts and loads the context info provided by Dagster. While the process runs, execution data, logs, and any specified metadata is streamed back to Dagster.

---
title: Dagster Pipes | Dagster Docs
description: "Dagster Pipes provides a protocol between the orchestration environment (Dagster) and external execution (ex: Databricks) and a toolkit for building implementations of that protocol."
---

# Dagster Pipes

Dagster Pipes is a toolkit for building integrations between Dagster and external execution environments. It standardizes the process of passing parameters, injecting context information, ingesting logs, and collecting metadata all while remaining agnostic to how remote computations are launched in those environments. This enables the separation of orchestration and business logic in the Dagster ecosystem.

> It also smooths the process of incorporating pre-existing code, business logic, and execution environments into Dagster. Using Dagster Pipes, you add just a few lines to your code and it becomes seamlessly invocable by the orchestrator. In turn, you can stream logs and metadata back to Dagster so you can leverage its observability, lineage, cataloging, and debugging capabilities.

Reviewer 4: I would remove the word `just` here, even a few lines of code can be more difficult than it appears. 
"it becomes seamlessly invocable by the orchestrator" I find this a little confusing as well, maybe we can simplify this whole sentence:

> With Dagster Pipes and a few lines of code, you can execute your code through the orchestrator. In turn, you can stream logs ... "

---

## Benefits

With Dagster Pipes, you can:

- Incorporate existing code into Dagster without huge refactors
- Onboard stakeholder teams onto Dagster incrementally
- Run code in external environments, and:
  - Easily pass parameters to the code
  - Stream unstructured logs and structured metadata back to Dagster
- Separate orchestration and business logic environments
- Use languages other than Python with Dagster

---

## Limitations

While Dagster Pipes is lightweight and flexible, there are a few limitations to be aware of:

> - **Step launchers and Pipes can't currently be used together.** Dagster Pipes is a lightweight alternative to step launchers. Think of this as an either-or situation - you can use step launchers **or** you can use Dagster Pipes, depending on your needs.

Reviewer 4: How do I decide what needs fit where? This tells me they don't work together but not how to decide between the two.

Author: This is a good call out - I struggled with this. Took a swing and updated this to:

> **Step launchers and Pipes can't currently be used together.** Dagster Pipes is a lightweight alternative to step launchers. Think of this as an either-or situation - you can use step launchers **or** you can use Dagster Pipes. If your external code requires the core Dagster module (see below), you should use step launchers instead of Pipes.

- **Some Dagster concepts aren't supported for use in external processes.** Dagster Pipes (`dagster-pipes`) doesn't include the core Dagster (`dagster`) library. As such, concepts like resources and I/O managers, which are included in `dagster`, aren't available for use in external processes executed by Dagster Pipes.

  For example, I/O managers aren't currently supported, as Dagster Pipes isn't built for processing data in memory. To process data in memory in Pyspark, you could use an I/O manager [as demonstrated in the Pyspark integration guide](/integrations/spark#running-pyspark-code-in-assets). Otherwise, in a remote case, you can use Pipes.

---

## How it works

Dagster Pipes provides a protocol between the orchestration environment (Dagster) and external execution (ex: Databricks) and a toolkit for building implementations of that protocol.

> When Dagster Pipes is invoked, several steps will be carried out both in the **orchestration process**, or Dagster, and in the **external process**, such as Databricks.

Reviewer 4: ```suggestion
When Dagster Pipes is invoked, several steps will be carried out in **Dagster's orchestration process** and in the **external process**, such as Databricks.
```

### In the orchestration process (Dagster)

> When Dagster Pipes is called in a Dagster asset, Dagster launches the external process with parameters and context information (ex: `partition_key`, `asset_key`, etc.)

Reviewer 4: Do we know upfront the exact parameters and context being passed? If so, it would be worth linking to the API here, otherwise I feel like I'm guessing
